controllers/routemap.go

```diff
@@ -0,0 +1,420 @@
+		// attempt to match status condition with config condition
+		for _, conditionUncast := range conditions {
+			condition, ok := conditionUncast.(map[string]interface{})
+			if !ok {
+				log.Logger.Info("unable to cast condition to map[string]interface{}")
+				return false
+			}
+			name, found, err := unstructured.NestedString(condition, "type")
+			if !found || err != nil || !strings.EqualFold(name, c.Name) {
+				log.Logger.Trace("condition with no type")
+				continue
+			}
+			status, found, err := unstructured.NestedString(condition, "status")
+			if !found || err != nil {
+				log.Logger.Trace("condition with no status")
+				continue
+			}
+
+			// found a match between config condition and status condition
+			generation, found, _ := unstructured.NestedInt64(u.Object, "metadata", "generation")
+			if found {
+				observedGeneration, found := getObservedGeneration(u, condition)
+				if found && observedGeneration < generation {
+					// condition generation does not equal resource generation
+					log.Logger.Trace("condition not satisfied")
+					return false
+				}
+			}
+
+			// check if condition status equals the required value specified in config
+			if !strings.EqualFold(status, c.Status) {
+				log.Logger.Info("condition not satisfied")
+				return false
+			}
+			satisfied = true
+		}
+		if !satisfied {
+			// this condition is still not satisfied
+			return false
+		}
+	}
+	return true
+}
+
+// getObservedGeneration attempts to get the observed generation value from a condition field
+// this is best effort and assumes api conventions are followed
+// https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties
+// this function is derived from:
+// https://github.com/kubernetes/kubectl/blob/master/pkg/cmd/wait/wait.go
+func getObservedGeneration(obj *unstructured.Unstructured, condition map[string]interface{}) (int64, bool) {
+	conditionObservedGeneration, found, _ := unstructured.NestedInt64(condition, "observedGeneration")
+	if found {
+		return conditionObservedGeneration, true
+	}
+	statusObservedGeneration, found, _ := unstructured.NestedInt64(obj.Object, "status", "observedGeneration")
+	return statusObservedGeneration, found
+}
+
+// validate routemap CM
+func validateRoutemapCM(confMap *corev1.ConfigMap) error {
+	// routemap CM must be immutable
+	if confMap.Immutable == nil || !(*confMap.Immutable) {
+		err := errors.New("routemap CM is not immutable")
+		log.Logger.Error(err)
+		return err
+	}
+	return nil
+}
+
+// validateRoutemap validates a given sbject
+func validateRoutemap(s *routemap, config *Config) (*routemap, error) {
+	// routemap must have at least one variant
+	if len(s.Variants) == 0 {
+		e := errors.New("routemap must at least one variant")
+		log.Logger.Error(e)
+		return nil, e
+	}
+
+	// if !clusterScoped, variant resource namespace should be nil or equal routemap namespace
```

Comment:
And if it is cluster scoped shouldn't we be checking that a namespace is always set?